### PR TITLE
Loans: Runtime API for updating the portfolio with fake prices

### PR DESCRIPTION
**Pull request by lemunozm**
# Description

Add a Runtime API to emulate a portfolio update with custom prices.

### CODE REVIEW

**Comment by lemunozm** [comment]
Blocked by https://github.com/centrifuge/centrifuge-chain/pull/1739

File: pallets/loans/src/lib.rs
Comment (mustermeiszer) [inline pallets/loans/src/lib.rs:1084]: What is the benfit of using the `PriceCollectionInput` here vs just using a `BTreeMap` as the parameter?

Comment (lemunozm) [reply pallets/loans/src/lib.rs:1084]: You can see each use case here: https://github.com/centrifuge/centrifuge-chain/pull/1748/files#diff-efd52f7fa525763070479bd9d72c66286451a6752bed8534e83e1004e755c3e1R469

Comment (lemunozm) [reply pallets/loans/src/lib.rs:1084]: `Empty` is just sugar for `Custom([])` and could be removed. But the caller should choose between `Custom` and `FromRegistry`.

Comment (wischli) [reply pallets/loans/src/lib.rs:1084]: I prefer @lemunozm's proposal over just using a `BTreeMap` as it is more explicit which is always helpful for runtime API which can be expected to be used by externals as well.

Comment (mustermeiszer) [reply pallets/loans/src/lib.rs:1084]: Okay, you changed the valuation parameters of the runtime-api. Sounds good!

 The most important thing though would be to mimic closing an epoch with 
* faked prices
* Optional<Solution>

Returning:
* Fullfillment amounts for each tranche for both invest and redeem
* New pool-reserve
* New Tranche prices

File: pallets/loans/src/lib.rs
Comment (wischli) [inline pallets/loans/src/lib.rs:1239]: A quick comprehension question: Our rather newer NAV calculation runtime API calls `Loans::update_nav(pool_id)` such that the Pool Fees NAV calculation can be based on that. Say the loans NAV is updated with a price collection input in block `i` and in the next block `i+1` we are querying the NAV via `PoolsApi::nav(pool_id)` which performs `Loans::update_nav(pool_id)` with an update based on `PriceCollectionInput::FromRegistry`. IIUC, the prices from the update in block `i` are still cached. What I want to  find out is if `Loans::update_nav` can make the Loans NAV less accurate if it was previously based on a custom price list.

**Review by wischli — COMMENTED** [review]
Really clean runtime API! I have a question before I hit the approval button.

File: pallets/loans/src/lib.rs
Comment (lemunozm) [reply pallets/loans/src/lib.rs:1084]: Ok, I guess somehow this should be amalgamated with some other `pallet-pool` calls to get this, right?

Then I propose to create another in the PoolSystem RuntimeAPI, maybe called simulate_close_epoch or similar, where all these calls are performed to return all the required values above.

File: pallets/loans/src/lib.rs
Comment (lemunozm) [reply pallets/loans/src/lib.rs:1239]: It was previously based on the price registry, so both lines above are exactly identical. From a pool fees perspective, nothing has changed. The only open door (right now) is for the RuntimeAPI.

File: pallets/loans/src/lib.rs
Comment (lemunozm) [reply pallets/loans/src/lib.rs:1084]: Great, I'll do in another PR pool-related.

**Review by mustermeiszer — APPROVED** [review]
LGTM! Can you ping apps that the runtime api changed?

**Review by wischli — COMMENTED** [review]
One minor nit change I would like to see: Runtime API can be versioned via `#[api_version($VERSION)]` as part of the `decl_runtime_apis!` macro ([example](https://github.com/KILTprotocol/kilt-node/blob/62a1373d20cae10a72e67973f5f6392052a6eaee/runtime-api/did/src/lib.rs#L69C2-L69C19)). However, we have not made use of this feature so far. Of course, we should to make the breaking changes explicit.

**Comment by lemunozm** [comment]
Good point. Just one question: should we bump up the version even if the upgrade is retro-compatible? (it means just an addition)

**Comment by lemunozm** [comment]
@mustermeiszer can we merge https://github.com/centrifuge/centrifuge-chain/pull/1739 before?

**Review by mustermeiszer — DISMISSED** [review]

**Comment by wischli** [comment]
> Good point. Just one question: should we bump up the version even if the upgrade is retro-compatible? (it means just an addition)

Definitely not a blocker because it's not breaking. In the future, I would still like to make real use of versioning because it can be tied to the earliest supported spec version for that API. 

For instance, SubQuery recently had issues with the new PoolsNAV API because it was attempted at blocks which didn't support it. But in order for this to fully work, Apps needs to add spec version granularity to the type and RtApi decorations.

**Review by wischli — DISMISSED** [review]

**Comment by lemunozm** [comment]
Good reasoning, you've convince me! => [24adcea](https://github.com/centrifuge/centrifuge-chain/pull/1748/commits/24adcea0fe32182aa7b12eadd6114b9513280247)

**Review by wischli — APPROVED** [review]

**Pull request merged**
